pkg/icingadb/history/sync.go

```diff
@@ -32,274 +35,366 @@ func NewSync(db *icingadb.DB, redis *icingaredis.Client, logger *zap.SugaredLogg
+					st := state[e]
+					delete(state, e)
+					stateMu.Unlock()
+
+					st.Pending--
+					if st.Pending == 0 {
+						select {
+						case out <- st.Message:
+						case <-ctx.Done():
+							return ctx.Err()
 						}
+					}
 
-						ids = nil
-						timeout = nil
+				case m, ok := <-skipped:
+					if !ok {
+						return nil
 					}
-				}
-			case <-timeout:
-				cmd := s.redis.XDel(ctx, stream, ids...)
-				if _, err := cmd.Result(); err != nil {
-					return icingaredis.WrapCmdErr(cmd)
-				}
 
-				ids = nil
-				timeout = nil
+					select {
+					case out <- m:
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+
+				case <-ctx.Done():
+					return ctx.Err()
+				}
 			}
-		}
+		})
+
+		return g.Wait()
 	}
 }
 
-// historyStream represents a Redis history stream.
-type historyStream struct {
-	// kind specifies the stream's purpose.
-	kind string
-	// structifiers lists the factories of the model structs the stream data shall be copied to.
-	structifiers []structify.MapStructifier
-}
+// userNotificationStage is a specialized stageFunc that populates the user_notification_history table. It is executed
+// on the notification history stream and uses the users_notified_ids attribute to create an entry in the
+// user_notification_history relation table for each user ID.
+func userNotificationStage(ctx context.Context, s Sync, key string, in <-chan redis.XMessage, out chan<- redis.XMessage) error {
+	type NotificationHistory struct {
+		Id            types.UUID   `structify:"id"`
+		EnvironmentId types.Binary `structify:"environment_id"`
+		EndpointId    types.Binary `structify:"endpoint_id"`
+		UserIds       string       `structify:"users_notified_ids"`
+	}
+
+	structifier := structify.MakeMapStructifier(reflect.TypeOf((*NotificationHistory)(nil)).Elem(), "structify")
 
-// historyStreams contains all Redis history streams to sync.
-var historyStreams = func() []historyStream {
-	var streams []historyStream
-	for _, rhs := range []struct {
-		kind       string
-		structPtrs []v1.UpserterEntity
-	}{
-		{"notification", []v1.UpserterEntity{(*v1.NotificationHistory)(nil), (*v1.HistoryNotification)(nil)}},
-		{"usernotification", []v1.UpserterEntity{(*v1.UserNotificationHistory)(nil)}},
-		{"state", []v1.UpserterEntity{(*v1.StateHistory)(nil), (*v1.HistoryState)(nil)}},
-		{"downtime", []v1.UpserterEntity{(*v1.DowntimeHistory)(nil), (*v1.HistoryDowntime)(nil)}},
-		{"comment", []v1.UpserterEntity{(*v1.CommentHistory)(nil), (*v1.HistoryComment)(nil)}},
-		{"flapping", []v1.UpserterEntity{(*v1.FlappingHistory)(nil), (*v1.HistoryFlapping)(nil)}},
-		{"acknowledgement", []v1.UpserterEntity{(*v1.AcknowledgementHistory)(nil), (*v1.HistoryAck)(nil)}},
-	} {
-		var structifiers []structify.MapStructifier
-		for _, structPtr := range rhs.structPtrs {
-			structifiers = append(structifiers, structify.MakeMapStructifier(reflect.TypeOf(structPtr).Elem(), "json"))
+	baseUUID := uuid.MustParse("235396c3-a525-4806-a3bc-3806c25baaeb")
```

Comment:
A namespace inside a table? Do you fear a collision with the migration tool?

Reply:
My idea was to avoid changing the schema here for this PR as we will likely change this when addressing HA + history. If keeping an explicit ID column, I'd have chosen `SHA1(env_ud, notification_history_id, user_id)` but the column is only 16 bytes, so a hash-based UUID is probably the closest I can get. Another option would be to drop the ID column and use a composite primary key over all remaining columns as this is a relation table and the relation is what identifies the row.

But all in all, I'd postpone the final ID choice to the HA + history task and see this as an intermediate solution.